Feature/add ufs s2s support

slawrence87544 opened the pull request:
Adding UFS support for AWS
fixes #52

Commit:
feat: modify config_custom.sh and stack_custom.yaml to suppor ufs-s2s

Commit:
refactor: modified stack_custum to add a few extra libraries

Commit:
fix: guarantee environment variables are set for shared=NO config

body: in build_pnetcdf.sh, if the shared config option is NO,
the configuration step will blow up due to an undefined
environment variable 'shared_flags'.  Likewise, if pnetcdf is
enabled for build_pio.sh, the PNETCDF env var might be undefined
if a default value is not set.  I also modified the
stack_custom.yaml file to disable the 'shared' option throughout.

Commit:
feat: adding stack config for the ufs_s2s_model use case

body: the main difference here is the addition of jpeg to support
the jasper build.

Commit:
feat: add container support for intel oneapi and ufs-s2s-model

body: This commit is meant to add ubuntu-based docker containers
which utilize Intel's license free compiler OneAPI along with
the latest hpc-stack.  The container support of the ufs-s2s-model
is provided as a multi-stage build such that one could re-use
the base containers for other purposes if they wish.  I have also
provided a couple readme files which help show how to setup
singularity (which can be run on-prem at least on Hera and in the
cloud) without root privilages.  Finally, some instructions on
how to convert docker images to singularity images are also
provided.

Commit:
Merge branch 'develop' into feature/add-ufs-s2s-support

aerorahul commented on config/config_custom.sh:
```diff
@@ -16,3 +16,4 @@ export   MAKE_CLEAN=N
 export DOWNLOAD_ONLY=N
 export STACK_EXIT_ON_FAIL=Y
 export WGET="wget -nv"
+
```
un-necessary

slawrence87544 replied:
Sorry, the extra line was unintentional.

aerorahul commented on config/stack_ufs_s2s.yaml:
```diff
@@ -0,0 +1,242 @@
+cmake:
```
Do we need `stack_ufs_s2s.yaml` and `stack_ufs_s2s_model.yaml` both?
How do they differ from `stack_noaa.yaml` or `stack_custom.yaml`.

slawrence87544 replied:
I had thought I had removed stack_ufs_s2s.yaml.  The main difference between stack_noaa.yaml and the stack_ufs_s2s_model.yaml are that it doesn't build eigen, ecbuild, eckit, fckit, or atlas as none of those are required.  I was trying to minimize the size of the subsequent container.

aerorahul commented on containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:
```diff
@@ -0,0 +1,47 @@
+ENV PNETCDF=/usr/local
+ENV NETCDF=/usr/local
+
+RUN git clone https://github.com/slawrence87544/hpc-stack.git && \
```
this would become `github.com/noaa-emc`

slawrence87544 replied:
Yes, I'll change that now.  It was only in there so that if someone wanted to try this out prior to the PR being accepted, they could actually do so.  Did you have a preference on how to handle that or is it assumed that people won't try to use this until the changes are merged?

aerorahul commented on containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:
```diff
@@ -0,0 +1,47 @@
+RUN git clone https://github.com/slawrence87544/hpc-stack.git && \
+    cd hpc-stack && \
+    git checkout feature/add-ufs-s2s-support && \
+    ./build_stack.sh -p /usr/local -c config/config_linux_ubuntu20_intel_oneapi_beta09.sh -y config/stack_ufs_s2s_model.yaml
```
Can we build this as `builder` instead in `/usr/local`?

aerorahul commented on containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:
```diff
@@ -0,0 +1,47 @@
+    rm cmake-3.16.4-Linux-x86_64.tar.gz && \
+    echo export PATH=/usr/local/cmake/bin:\$PATH >> /etc/bashrc
+
+ENV PNETCDF=/usr/local
+ENV NETCDF=/usr/local
```
why are these necessary here?

slawrence87544 replied:
They may not be necessary, but I was having trouble building some of the packages without them.  Perhaps there is a better way to ensure that those environment variables get created and exported along with NETCDF_ROOT/PNETCDF_ROOT.

aerorahul commented on containers/how-to-create-singularity-image-from-local-docker-image.txt:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
If this is a `.txt` why the `bash` interpreter?   Also, this reads more like a script than a how-to.

aerorahul commented on containers/Dockerfile.intel-oneapi-ufs-s2s-coupled-lite:
```diff
@@ -0,0 +1,14 @@
+FROM intel-oneapi-ufs-s2s-coupled-dev:ubuntu20 as ufs
```
what is the use case of this dockerfile?

slawrence87544 replied:
In order to reduce the size of the containers, it's common to delete all the source code, but if one simply removes it in the same container image where it was created, it doesn't reduce the size (hence the need for a multi-stage container build).  It really depends on how one plans to use the container, but a smaller container is easier to move around in general and loads up faster when bringing up a cloud-based master instance.

aerorahul commented on libs/build_pio.sh:
```diff
@@ -32,6 +32,7 @@ if $MODULES; then
     fi
 else
     prefix=${PIO_ROOT:-"/usr/local"}
+    PNETCDF=${PNETCDF:-"/usr/local"}
```
This is not correct.

slawrence87544 replied:
I encountered an issue when building the stack with pnetcdf enabled.  I had PNETCDF defined outside of this script, but when I ran the script, it blew up complaining about an undefined PNETCDF.  Maybe it was user error.  I'll just remove it.

aerorahul reviewed (changes requested):
little comments here and there.

aerorahul commented on containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:
```diff
@@ -0,0 +1,47 @@
+    cd hpc-stack && \
+    git checkout feature/add-ufs-s2s-support && \
+    ./build_stack.sh -p /usr/local -c config/config_linux_ubuntu20_intel_oneapi_beta09.sh -y config/stack_ufs_s2s_model.yaml
+
```
Please `rm -rf hpc-stack` after building to lighten up the container. There is no need to retain the stack and build directories of the software.

slawrence87544 replied:
Got it.  I'll remove the src code.  I didn't know if the esmf library needed to preserve it's source code in order to be functional.

Commit:
fix: address PR comments from Rahul Mahajan

body: Specifically install software in /home/builder instead of
in /usr/local.  Remove the step of setting environment variables
PNETCDF and NETCDF in container definition file for building the
hpc-stack.  Removed the bash interpreter in the singularity install
instructions.  Removed unnecessary stack config file.

aerorahul commented on containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:
```diff
@@ -24,22 +24,19 @@ ENV I_MPI_LIBRARY_KIND='release_mt'
     curl -f -s -S -R -L -O https://github.com/Kitware/CMake/releases/download/v3.16.4/cmake-3.16.4-Linux-x86_64.tar.gz && \
-    mkdir -p /usr/local/cmake && \
-    tar zxf cmake-3.16.4-Linux-x86_64.tar.gz -C /usr/local/cmake --strip-components=1 && \
+    mkdir -p /usr/builder/cmake && \
```
I think it is `/home/builder` not `/usr/builder`

aerorahul commented on containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:
```diff
@@ -24,22 +24,19 @@ ENV I_MPI_LIBRARY_KIND='release_mt'
 
 RUN cd /tmp && \
```
Make sure to delete stuff from `/tmp` to decrease the image size.

aerorahul commented on containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:
```diff
@@ -24,22 +24,19 @@ ENV I_MPI_LIBRARY_KIND='release_mt'
-RUN echo 'export PATH=/usr/local/bin:$PATH'      >> /etc/bash.bashrc && \
-    echo 'export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH' >> /etc/bash.bashrc && \
-    echo 'export CMAKE_PREFIX_PATH=/usr/local'   >> /etc/bash.bashrc && \
+RUN echo 'export PATH=/usr/builder/bin:$PATH'      >> /etc/bash.bashrc && \
```
The user is `builder` and the space is `/home/builder`

aerorahul reviewed (changes requested):
The user `builder` path is `/home/builder` not `/usr/builder`

Commit:
Merge branch 'develop' into feature/add-ufs-s2s-support

Commit:
fix: disable pnetcdf in pio build on ufs_s2s_model stack config

Commit:
fix: replace /usr/builder with /home/builder

aerorahul commented on containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:
```diff
@@ -24,19 +24,19 @@ ENV I_MPI_LIBRARY_KIND='release_mt'
     cd hpc-stack && \
     git checkout feature/add-ufs-s2s-support && \
-    ./build_stack.sh -p /usr/builder -c config/config_linux_ubuntu20_intel_oneapi_beta09.sh -y config/stack_ufs_s2s_model.yaml
+    ./build_stack.sh -p /home/builder -c config/config_linux_ubuntu20_intel_oneapi_beta09.sh -y config/stack_ufs_s2s_model.yaml
```
Can you make the install prefix `/home/builder/opt` or something like that?
It is probably not a good idea to treat the home directory as an installation prefix.

aerorahul commented on containers/Dockerfile.intel-oneapi-hpc-stack-ubuntu20:
```diff
@@ -24,19 +24,19 @@ ENV I_MPI_LIBRARY_KIND='release_mt'
     curl -f -s -S -R -L -O https://github.com/Kitware/CMake/releases/download/v3.16.4/cmake-3.16.4-Linux-x86_64.tar.gz && \
-    mkdir -p /usr/builder/cmake && \
-    tar zxf cmake-3.16.4-Linux-x86_64.tar.gz -C /usr/builder/cmake --strip-components=1 && \
+    mkdir -p /home/builder/cmake && \
```
`/home/builder/opt/cmake`

Commit:
fix: modify containers to install software in /home/builder/opt

body: Also revert config/stack_noaa.yaml to what is in the develop
branch.  Slight modification to stack_ufs_s2s_model.yaml to enable
szip.

Commit:
fix: revise method of setting intel oneapi env vars

body: Remove docker build logic to replace sh/dash shell with bash.
Also remove logic to set symlink for gmake.

edwardhartnett commented:
Is there an issue for this work? If so, please refer to it in the PR. If not, please create one. ;-)

Commit:
fix: for intel containers  all software is installed at /home/builder/opt

body: In response to comments on the draft PR, we have moved all intel
container software installs to /home/builder/opt with the exception of
the intel install which by default gets installed at /opt/intel.

Commit:
refactor: remove hpc-stack src code and build directories

edwardhartnett commented:
@slawrence87544 is there an issue for this work? If not please add one. If so, please reference it in your PR description. 

Also please always add a PR description for any PRs to this repo.

Commit:
refactor: disable udunits for stack_mac.yaml

body: as per Rahul's suggestion, I'm disabling udunits compile
for the stack_mac.yaml to avoid the mpich/clang compile error.

I also removed an unnecessary extra line in config_custom.sh that
I inadvertently introduced.

aerorahul commented on containers/Dockerfile.intel-oneapi-ufs-s2s-coupled-dev:
```diff
@@ -0,0 +1,35 @@
+
+RUN apt-get update -y && \
+    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends -o=Dpkg::Use-Pty=0 \
+        csh \
```
Do we need all these shells? `csh`, `ksh`, `tcsh`?
Also, why have `tex, lynx, nedit, emacs`? Is `vi` not enough?
We are not building documentation, so `doxygen` is not necessary either.

slawrence87544 replied:
These extra packages were included based on a jcsda container found here (https://github.com/JCSDA-internal/containers/blob/develop/Dockerfile.intel19-impi-dev).  Since I don't have much experience with how these dev containers will be used, I put in more shell options as I guessed that researchers might want them to be available.  I will remove them as they are not critical to the ufs-s2s-model's functionality.

aerorahul commented on containers/how-to-create-singularity-image-from-local-docker-image.txt:
```diff
@@ -0,0 +1,139 @@
+# This guide assumes that docker is already installed
```
Can you make this a markdown.  The formatting is so much better.  I can help.

aerorahul reviewed (changes requested):
some minor comments.
getting closer to finished.

aerorahul commented on config/stack_ufs_s2s_model.yaml:
```diff
@@ -0,0 +1,251 @@
+cmake:
```
I can check, but how different is this from `stack_noaa.yaml`

slawrence87544 replied:
The only differences are inclusion of the jpeg compile and the inclusion of szip functionality for the hdf5 compile.  I can remove them and see how the compile goes.  If it still works, I'll simply remove that config.

aerorahul replied:
jpeg is OFF in stack_noaa.yaml, but we intend to turn it ON.
The UFS uses zlib for compression and that is why the szip is OFF.

aerorahul committed:
Merge branch 'develop' into feature/add-ufs-s2s-support

Commit:
refactor: remove unnecessary files and packages

body: As the stack_ufs_s2s_model.yaml was nearly identical to
stack_noaa.yaml, we switched over to using stack_noaa.yaml
for the hpc-stack container build for intel oneapi.  Also
removed unnecessary package installs within the ufs-s2s
container build.  Finally, translated singularity install
guide into markdown and renamed it to container_development.md

Commit:
Merge branch 'feature/add-ufs-s2s-support' of github.com:slawrence87544/hpc-stack into feature/add-ufs-s2s-support

aerorahul reviewed (approved):
Thanks @slawrence87544

Commit:
refactor: changed naming of two container files and removed file

body: The guide for how to transform a docker container into a
singularity container was made into a markdown document and renamed
as container-development.md

Commit:
feat: refer to official hpc-stack repo and checkout develop branch

body: Replace reference to my forked hpc-stack in the container
Dockerfile.intel-oneapi-hpc-stack-ubuntu20 with the official
repo and replace the temporary branch with the develop branch
in that same file.

Commit:
fix: update referenced docker image name

body: in the docke file Dockerfile.intel-oneapi-ufs-weather-model-lite
the referenced docker image name needed to be updated to match
the new expected name 'intel-oneapi-ufs-weather-model'.
